### PR TITLE
[ENH] Squaring residuals estimator

### PR DESCRIPTION
**Pull request by kcc-lion**
#### Reference Issues/PRs
Implements  #2698

#### What does this implement/fix? Explain your changes.
Implementation of Squaring Residuals estimator as specified in above issue.

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?

Still thinking about testing and implementation (so Draft PR).

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

##### For new estimators
- [x] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

File: sktime/forecasting/squaring_residuals.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+from sktime.forecasting.model_selection import ExpandingWindowSplitter
+from sktime.forecasting.naive import NaiveForecaster
+
+
+class SquaringResiduals(BaseForecaster):
+    r"""Compute the prediction variance based on a separate forecaster.
+
+    Wraps a `forecaster` with another `variance_forecaster` object that
+    allows for quantile and interval estimation by fitting the
+    `variance_forecaster` to the rolling residuals.
```
Comment (fkiraly) [inline sktime/forecasting/squaring_residuals.py:23]: I would use "volatility" or "residual", since variance has a specific meaning (of the residual distribrution) that does not apply for all choices of h

File: sktime/forecasting/squaring_residuals.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+    Let :math:`t_1, \dots, t_N` be the train set.
+    Let `steps_ahead` be a positive integer indicating the steps ahead
+    we want to forecast the residuals. Let `initial_window` be
+    the minimal number of observations to which the forecaster is fitted.
+
+    1. For :math:`i = initial\_window, \dots, N - steps\_ahead`
+        a. Train/Update forecaster A on :math:`y(t_1), \dots, y(t_i)`
+        b. Make point prediction for :math:`t_{i+steps\_ahead}` to get
+           :math:`\hat{y}(t_{i+steps\_ahead})`
+        c. Compute the residual for :math:`t_{i+steps\_ahead}` with
```
Comment (fkiraly) [inline sktime/forecasting/squaring_residuals.py:35]: with -> as

File: sktime/forecasting/squaring_residuals.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+           :math:`r(t_{i+steps\_ahead}) := y(t_{i+steps\_ahead})
+           - \hat{y}(t_{i+steps\_ahead})`
+        d. Apply  :math:`e(t_{i+steps\_ahead}) := h(r(t_{i+steps\_ahead}))`
+           where :math:`h(x)` is given by :math:`strategy`
```
Comment (fkiraly) [inline sktime/forecasting/squaring_residuals.py:39]: "apply" is not a clear instruction -> compute

File: sktime/forecasting/squaring_residuals.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/squaring_residuals.py:23]: or "scale", that´s another option

File: sktime/forecasting/squaring_residuals.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    Prediction for :math:`t_{N+steps\_ahead}` is done as follows:
+
+    1. Use `forecaster` to predict location param :math:`\hat{y}(t_{N+steps\_ahead})`
+    2. Use `variance_forecaster` to predict scale param :math:`e(t_{N+steps\_ahead})`
+    3. Calculate prediction intervals based on e.g. normal assumption
+       :math:`N(\hat{y}(t_{N+steps\_ahead}),  e(t_{N+steps\_ahead}))`
+
+    Parameters
+    ----------
+    forecaster : sktime.estimator, optional
```
Comment (fkiraly) [inline sktime/forecasting/squaring_residuals.py:52]: `sktime.estimator` is incorrect (there is no such module)

suggest:
`sktime` forecaster, `BaseForecaster` descendant

File: sktime/forecasting/squaring_residuals.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+    initial_window : int, optional, default=2
+        Size of initial_window to which forecaster is fitted
+    steps_ahead : int, optional, default=1
+        Steps ahead for which we predict the residuals
+    strategy : str, optional, default='square'
+        Function applied to the residuals
+    distr : str, optional, default='norm'
+        Distributional assumption (["norm", "laplace", "t", "cauchy"])
+    distr_kwargs : dict, optional
+        Additional arguments required by the distributional assumption
```
Comment (fkiraly) [inline sktime/forecasting/squaring_residuals.py:65]: ... by the distribution

File: sktime/forecasting/squaring_residuals.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    >>> sqr = sqr.fit(y)
+    >>> fh = ForecastingHorizon(values=[1, 2, 3])
+    >>> pred_interval = sqr.predict_interval(fh, coverage=0.95)
+    """
+
+    _tags = {
+        "scitype:y": "univariate",  # which y are fine? univariate/multivariate/both
+        "ignores-exogeneous-X": True,  # does estimator ignore the exogeneous X?
+        "handles-missing-data": False,  # can estimator handle missing data?
+        "y_inner_mtype": "pd.Series",  # which types do _fit, _predict, assume for y?
```
Comment (fkiraly) [inline sktime/forecasting/squaring_residuals.py:87]: why can´t we do this multivariate? Should not be too hard, use `pd.DataFrame` here.
(maybe I don´t see a problem though)

File: sktime/forecasting/squaring_residuals.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+class SquaringResiduals(BaseForecaster):
+    r"""Compute the prediction variance based on a separate forecaster.
+
+    Wraps a `forecaster` with another `variance_forecaster` object that
+    allows for quantile and interval estimation by fitting the
+    `variance_forecaster` to the rolling residuals.
+
+    Fitting proceeds as follows:
+    Let :math:`t_1, \dots, t_N` be the train set.
+    Let `steps_ahead` be a positive integer indicating the steps ahead
```
Comment (fkiraly) [inline sktime/forecasting/squaring_residuals.py:27]: this should be `fh`, not a parameter

File: sktime/forecasting/squaring_residuals.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+        Returns
+        -------
+        y_pred : pd.Series
+            Point predictions
+        """
+        fh_abs = fh.to_absolute(self.cutoff)
+        y_pred = self._forecaster_.predict(X=X, fh=fh_abs)
+        return y_pred
+
+    def _update_predict_single(self, y, fh, X=None, update_params=True):
```
Comment (fkiraly) [inline sktime/forecasting/squaring_residuals.py:206]: why do we override this? is the default not identical to this?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks great!

One blocking change relates to the `steps_ahead` - should that not just be `fh`, and we loop over `fh` elements internally?

Other comments relate to docstrings mostly.

**Review by fkiraly — APPROVED** [review]
Looks great, thanks!

Non-blocking change requests:
* `self._res_step_forecaster` is never read and overwritten repeatedly by the loop. This should just be a temp variable, not an object variable?
* there should be at least a second test parameter set, optimally covering different general cases or important edge cases

**Review by fkiraly — APPROVED** [review]
Great, thanks!
